src/v/storage/mvlog/versioned_log.cc

```diff
@@ -0,0 +1,173 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "storage/mvlog/versioned_log.h"
+
+#include "container/fragmented_vector.h"
+#include "model/offset_interval.h"
+#include "storage/mvlog/file.h"
+#include "storage/mvlog/logger.h"
+#include "storage/mvlog/readable_segment.h"
+#include "storage/mvlog/segment_appender.h"
+
+#include <seastar/core/circular_buffer.hh>
+#include <seastar/core/io_priority_class.hh>
+#include <seastar/core/lowres_clock.hh>
+
+#include <chrono>
+
+using namespace std::literals::chrono_literals;
+
+namespace storage::experimental::mvlog {
+
+active_segment::active_segment(
+  std::unique_ptr<file> f, model::offset o, segment_id id, size_t target_size)
+  : segment_file(std::move(f))
+  , appender(std::make_unique<segment_appender>(segment_file.get()))
+  , readable_seg(std::make_unique<readable_segment>(segment_file.get()))
+  , construct_time(ss::lowres_clock::now())
+  , target_max_size(target_size)
+  , id(id)
+  , base_offset(o)
+  , next_offset(o) {}
+
+active_segment::~active_segment() = default;
+
+readonly_segment::~readonly_segment() = default;
+readonly_segment::readonly_segment(std::unique_ptr<active_segment> active_seg)
+  : segment_file(std::move(active_seg->segment_file))
+  , readable_seg(std::move(active_seg->readable_seg))
+  , id(active_seg->id)
+  , offsets(model::bounded_offset_interval::checked(
+      active_seg->base_offset, model::prev_offset(active_seg->next_offset))) {}
+
+versioned_log::versioned_log(storage::ntp_config cfg)
+  : ntp_cfg_(std::move(cfg)) {}
+
+ss::future<> versioned_log::roll_unlocked() {
+    vassert(
+      !active_segment_lock_.ready(),
+      "roll_unlocked() must be called with active segment lock held");
+    vassert(active_seg_ != nullptr, "Expected an active segment");
+    vlog(
+      log.info,
+      "Rolling segment file {}",
+      active_seg_->segment_file->filepath().c_str());
+    co_await active_seg_->segment_file->flush();
+    auto ro_seg = std::make_unique<readonly_segment>(std::move(active_seg_));
+    segs_.emplace_back(std::move(ro_seg));
+}
+
+ss::future<> versioned_log::create_unlocked(model::offset base) {
+    vassert(
+      !active_segment_lock_.ready(),
+      "create_unlocked() must be called with active segment lock held");
+    vassert(active_seg_ == nullptr, "Expected no active segment");
+    const auto new_path = fmt::format(
+      "{}/{}.log", ntp_cfg_.work_directory(), next_segment_id_());
+    vlog(log.info, "Creating new segment file {}", new_path);
+    auto f = co_await file_mgr_.create_file(std::filesystem::path{new_path});
+    auto active_seg = std::make_unique<active_segment>(
+      std::move(f), base, next_segment_id_, compute_max_segment_size());
+    active_seg_ = std::move(active_seg);
+    ++next_segment_id_;
+}
+
+ss::future<> versioned_log::apply_segment_ms() {
+    auto lock = active_segment_lock_.get_units();
```

Comment:
missing co_await? i thought this would fail 🤔 or we never added this diagnostic? https://github.com/llvm/llvm-project/pull/76101

Reply:
Yikes! Fixed

Reply:
Added a test for this as well.